feat: introduce `DatabaseRole` CRD for declarative role management

jsilvela opened the pull request:
Introduce a `DatabaseRole` custom resource to manage a PostgreSQL role as a standalone Kubernetes object, instead of declaring it inline in the Cluster's `.spec.managed.roles` stanza. Each role gets its own lifecycle, status, and RBAC, which suits GitOps workflows and lets role definitions live next to the applications that own them. The spec reuses the same `RoleConfiguration` structure as the inline method, so migrating a role is a matter of moving the stanza into its own manifest.

Reconciliation follows the split already used by the `Database`, `Publication`, and `Subscription` resources: the instance manager applies the role to PostgreSQL and reports the result in the resource status, while the operator watches the password secrets and relays their changes through a status condition, so the instance manager never needs to watch secrets across the namespace.

The `databaseRoleReclaimPolicy` field controls what happens to the role when the resource is deleted, `retain` (default) or `delete`, mirroring persistent volumes. A role is dropped only by the object that actually reconciled it, and `ensure: absent` is rejected in favor of deleting the resource. When the same role name is managed both inline and through a `DatabaseRole`, the Cluster spec takes precedence and the resource reports the conflict instead of fighting over it.

The inline `managed.roles` stanza remains fully supported; the `DatabaseRole` resource is the recommended method going forward.

Closes #5940
Closes #5341
Closes #5967

Assisted-by: Claude Opus 4.8

jsilvela committed:
feat: add Role CRD

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
refactor: make the role reconciler code more reusable

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
chore: reconcile CRD Roles

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
fix: add the missing permissions logic and complete the role reconciler

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
docs: add example roles YAML

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
fix: rebase errors

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gabriele-wolfox committed:
chore: review of refactor code: naming

Signed-off-by: wolfox <gabriele.quaresima@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gabriele-wolfox committed:
chore: review of role CRD

Signed-off-by: wolfox <gabriele.quaresima@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
fix: eliminate stale dependency

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
chore: rename roles to pgroles for collision avoidance

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
chore: add the new yaml for pgroles

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela commented on internal/management/controller/roles/contract.go:
```diff
@@ -120,3 +123,37 @@ func (d *DatabaseRole) isEquivalentTo(inSpec apiv1.RoleConfiguration) bool {
 	return reflect.DeepEqual(role, spec) && d.hasSameValidUntilAs(inSpec)
 }
+
+// ApplyPassword updates a database role with the password located in the Secret
+// it returns the resource version of the Secret
```
```suggestion
// ApplyPassword updates a database role with the password located in the Secret,
// and it returns the resource version of the Secret
```

jsilvela commented on internal/management/controller/roles/contract.go:
```diff
@@ -120,3 +123,37 @@ func (d *DatabaseRole) isEquivalentTo(inSpec apiv1.RoleConfiguration) bool {
+	rolePassword passwordManager,
+	namespace string,
+) (string, error) {
+	var passVersion string
+	switch {
```
```suggestion
	switch {
```

jsilvela commented on internal/management/controller/roles/contract.go:
```diff
@@ -120,3 +123,37 @@ func (d *DatabaseRole) isEquivalentTo(inSpec apiv1.RoleConfiguration) bool {
+		}
+
+		d.Password = sql.NullString{Valid: true, String: passwordSecret.password}
+		passVersion = passwordSecret.version
+		return passVersion, nil
```
```suggestion
		return passwordSecret.version, nil
```

jsilvela commented on internal/management/controller/roles/contract.go:
```diff
@@ -120,3 +123,37 @@ func (d *DatabaseRole) isEquivalentTo(inSpec apiv1.RoleConfiguration) bool {
+		return "",
+			fmt.Errorf("cannot reconcile: password both provided and disabled: %s",
+				rolePassword.GetRoleSecretName())
+	default: // role.PasswordSecret != nil && !rolePassword.ShouldDisablePassword():
```
```suggestion
	default:
```

jsilvela reviewed (commented):
suggestions from the other PR https://github.com/cloudnative-pg/cloudnative-pg/pull/5973

NiccoloFei committed:
chore: review ApplyPassword

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela committed:
chore: update codegen, fix dictionary

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

jsilvela commented:
Thanks @sxd for reopening. Whew!

jsilvela commented:
@sxd ??? why has this been deleted again?

leonardoce committed:
chore: rename pgrole to role

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
fix: grant role permission to operator

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
feat: field indexer to associate Roles to clusters

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: reuse the RoleConfiguration structure

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: pgrole vs role

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: review

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce commented:
/test

gbartolini commented:
In here we decided to call the CRD `DatabaseRole` instead of `Role`: https://github.com/cloudnative-pg/cloudnative-pg/issues/5341#issuecomment-3766909464

I am sorry about that @leonardoce, is it possible to change it now or do you prefer to keep `Role`?

gbartolini committed:
docs: first part

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gbartolini committed:
docs: reorganised roles page

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gbartolini commented:
I have been playing with the feature and I love it. I have updated the docs (at the moment with the `Role` CRD name, which @leonardoce will replace with `DatabaseRole` once he's around). There is only one case that doesn't resonate with me.

If I create this role:

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Role
metadata:
  name: role-app
spec:
  cluster:
    name: cluster-example
  name: app
  createdb: true
  login: true
  roleReclaimPolicy: delete
```

Then, if I try to delete the role, the operation hangs (the finalizer is not able to successfully complete), the `status` of the role doesn't mention anything (and the role is not `Terminating`), the Postgres logs show:

```console
2026-04-04T19:43:30.792 ERROR    cluster-example-1 postgres         role "app" cannot be dropped because some objects depend on it
2026-04-04T19:43:30.792 ERROR    cluster-example-1 instance-manager Reconciler error
2026-04-04T19:48:58.479 ERROR    cluster-example-1 postgres         role "app" cannot be dropped because some objects depend on it
2026-04-04T19:48:58.479 ERROR    cluster-example-1 instance-manager Reconciler error
```

Ideally we should be able to see that a role is pending deletion and its status reflects the error (what we see in the Postgres logs). Is it possible to implement it like this @leonardoce? Thanks.

gbartolini committed:
docs: update operator capability levels

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gbartolini committed:
docs: use the new method with logical replication examples

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

gbartolini committed:
docs: cosmetic changes

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: Role -> DatabaseRole

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
fix: set status when deletion fails

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: fix linters

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: fix documentation links

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce commented:
/test

gbartolini committed:
docs: restored previous role management file name

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce committed:
chore: restore naming for examples

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

leonardoce commented:
/test

GellertBabel commented on internal/management/controller/databaserole_controller.go:
```diff
@@ -0,0 +1,362 @@
+	if err != nil {
+		return "", fmt.Errorf("while getting the role password: %w", err)
+	}
+	for _, r := range rolesInDB {
```
I think we are missing an if like
```
if role.Spec.Ensure == apiv1.EnsureAbsent {
				return "", roles.Delete(ctx, db, dbRole)
			}
```

mnencia replied:
Since this comment was written, the handling changed direction: `ensure: absent` is now rejected for `DatabaseRole`, both by a CEL rule on the CRD and by a runtime guard in the reconciler, with a message pointing the user to `roleReclaimPolicy: delete`. For a standalone resource, the reclaim policy is the natural way to express "drop the role when the object is removed", mirroring the PV/PVC behavior. The `ensure` field is still present in the spec only because the structure is shared with the inline `managed.roles` stanza.

GellertBabel commented on internal/management/controller/databaserole_controller.go:
```diff
@@ -0,0 +1,362 @@
+			return passwordVersion, roles.Update(ctx, db, dbRole)
+		}
+	}
+
```
Also when absent is true we should return nil?

```
if role.Spec.Ensure == apiv1.EnsureAbsent {
		return "", nil
	}
```

mnencia replied:
Same as the thread above: `ensure: absent` is rejected before reaching this code, so neither branch is needed here.

GellertBabel reviewed (dismissed):
Is ensure absent left out intentionally?

armru committed:
chore(databaserole): fix copy-paste artifacts and cleanup dead code

Cosmetic fixes across the DatabaseRole CRD, none affecting behavior:

- Fix godoc on SetAsFailed/SetAsApplied (said "publication"/"subscription")
- Fix ConditionPasswordSecretChange comment (described System ID check)
- Fix type description ("databaseroles API" not "databases API")
- Fix DatabaseRoleList comment ("DatabaseRoles" not "Roles")
- Fix misleading comment claiming no finalizers are managed
- Remove duplicate role.Status.Message assignment in failedReconciliation
- Remove unused Scheme field from instance-side DatabaseRoleReconciler
- Fix field indexer key to match actual JSON path (.spec.cluster.name)
- Fix E2E test timeout from 3000s to 300s (consistent with other tests)

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix(databaserole): fix validUntil reset and unblock retain-policy deletion

Two behavioral bugs in the instance-side DatabaseRole reconciler:

1. Removing validUntil from a DatabaseRole spec had no effect on the
   PostgreSQL role. The roleAdapter was always constructed with
   validUntilNullIsInfinity=false, so a nil ValidUntil produced no
   VALID UNTIL clause in the ALTER ROLE — PostgreSQL silently kept the
   old expiry. The managed-roles synchronizer correctly derives this
   flag from the existing DB state; this commit does the same.

2. GetSuperUserDB() was called unconditionally before the
   deletion/finalizer logic. When the database was unreachable, even
   the ReclaimRetain path (which only needs to remove the finalizer)
   was blocked. The DB connection is now obtained only inside the
   ReclaimDelete branch where it is actually needed.

Also fixes Role → DatabaseRole in operator_capability_levels.md and
adds the required cnpg.io/reload label to the sample Secret in
role-examples.yaml.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix(databaserole): add CEL validation and reserved-role guard

The CRD lacked admission-time validation that sibling CRDs (Database,
Subscription) already enforce. A user could target reserved system
roles (postgres, streaming_replica, pg_*, cnpg_*), mutate spec.name or
spec.cluster after creation (orphaning the PostgreSQL role on the old
cluster), or set ensure: absent (which was silently ignored by the
controller).

This commit:

- Adds a runtime guard in reconcileRole rejecting reserved role names
- Adds CEL XValidation rules on DatabaseRoleSpec for:
  - Immutability of spec.name and spec.cluster (prevents orphaned roles)
  - Reserved role name rejection
  - Rejection of ensure: absent (users should delete the K8s resource
    with roleReclaimPolicy: delete instead)
- Adds a runtime guard rejecting ensure: absent with a clear message
- Fixes kind: Role → DatabaseRole in logical_replication.md and
  cluster-example-logical-source.yaml

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix(databaserole): add passwordSecret/disablePassword CEL guard and clarify ensure docs

The passwordSecret and disablePassword fields were mutually exclusive
only at runtime — the CRD accepted both simultaneously, and the
reconciler failed with applied: false. The code comment explicitly
stated this should be prevented by a validation webhook, but none
existed.

Add a CEL XValidation rule on DatabaseRoleSpec rejecting the
combination at admission time, and update the contract.go comment to
reflect that CEL (not a webhook) now enforces this for DatabaseRole
CRDs.

Also clarify in declarative_role_management.md that ensure: absent is
only supported for inline managed roles. For DatabaseRole CRDs, the
equivalent is deleting the Kubernetes resource with
roleReclaimPolicy: delete.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix(databaserole): prevent perpetual reconciliation after password secret removal

Removing passwordSecret from a DatabaseRole that previously had one
caused a perpetual 30-second reconciliation loop. The operator-side
controller returns early when PasswordSecret is nil without clearing
the stale ConditionPasswordSecretChange condition. The instance-side
isAlreadyReconciled then perpetually finds a mismatch between the
empty SecretResourceVersion and the stale condition message, triggering
a full reconciliation (GetSuperUserDB + roles.List + roles.Update +
status patch) every 30 seconds for no effect.

Fix: skip the condition comparison in isAlreadyReconciled when no
password secret is configured. The generation check alone is
sufficient to detect spec changes in this case.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
test(databaserole): verify validUntil, comment, and attribute updates in E2E

The E2E tests for DatabaseRole verified create/delete/password but
never checked validUntil or comment in PostgreSQL (both set in
fixtures but unasserted), and never tested updating role attributes
after creation.

- Extend assertRoleHasExpectedFields to verify rolvaliduntil and
  shobj_description when the spec sets validUntil or comment
- Add "Role CR attributes are updated" test: creates a role, verifies
  initial attributes, patches login/createdb/connectionLimit, then
  verifies the updated values in PostgreSQL

These match the coverage depth of the existing managed_roles_test.go
for the inline path.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru commented:
/test

armru committed:
fix(databaserole): fix silent comment-update bug, add TransactionID tracking, and CEL/cleanup fixes

- Fix comment updates silently dropped for existing roles: roles.Update
  does not issue COMMENT ON ROLE, so call roles.UpdateComment when the
  comment differs from the existing DB value.

- Track TransactionID after role create/update by calling
  GetLastTransactionID and storing it in status.PasswordState for
  observability and parity with the inline RoleSynchronizer.

- Add CEL validation rule size(self.name) > 0 to reject empty role
  names at admission time rather than at PostgreSQL runtime.

- Clear stale PasswordSecretChange condition in operator-side controller
  when passwordSecret is removed from the spec.

- Remove dead GetRoleInherit() shadowing the promoted RoleConfiguration
  method, simplify crdRoleSecretName, and fix copy-paste comment
  artifacts.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
test(databaserole): extend E2E to cover comment/validUntil updates and fix SQL quoting

- Add validUntil and comment fields to the attribute-update E2E test so
  that changes to these fields after initial creation are verified in
  PostgreSQL, closing a coverage gap where the commit message claimed
  this coverage but the test did not exercise it.

- Add ObservedGeneration == Generation guard in the Eventually block to
  prevent false positives from stale status after a spec update.

- Replace '%s' string interpolation with pq.QuoteLiteral() in all SQL
  queries for consistency with assertInRoles.

- Reuse assertRoleHasExpectedFields helper in the attribute-update test
  instead of duplicating inline queries.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru commented:
/test

armru committed:
fix(databaserole): guard disablePassword CEL rule with has() check

Add !has(self.disablePassword) guard to the passwordSecret/disablePassword
mutual exclusivity CEL rule to prevent evaluation errors when
disablePassword is not set in the spec.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru commented:
/test

armru commented:
/test

jonasbadstuebner commented on config/crd/bases/postgresql.cnpg.io_databaseroles.yaml:
```diff
@@ -0,0 +1,291 @@
+                  "present"
+                enum:
+                - present
+                - absent
+                type: string
+              inRoles:
+                description: |-
+                  List of one or more existing roles to which this role will be
+                  immediately added as a new member. Default empty.
+                items:
+                  type: string
+                type: array
```
I still hope for the feature I proposed in #10007 to become a reality.
And would open a PR when the introduction of the `DatabaseRole` CRD is done.

But I want to prevent introducing a deprecated field to the CRD shortly after its release.

To allow for more flexibility with the grants, I suggest the following change to the CRD:
```suggestion
              roleGrants:
                description: |-
                  List of one or more existing roles to which this role will be
                  added as a new member. Default empty.
                items:
                  properties:
                    name:
                      description: The name of the role this role will be granted.
                      type: string
                    required:
                    - name
                  type: object
                type: array
```
This would allow me to reuse the `roleGrants` field and introduce `inherit`, `admin` and `set` as new properties to it, instead of creating a separate field where another migration would be necessary.  
This would make my PR less invasive and prevent a new code path from being created.

For users that don't want or need the feature from #10007 the change is trivial. They have to migrate to the `DatabaseRole` CRD anyway, so making small changes to the list of desired roles should not be a problem.
And they would not be impacted by my changes later, if the CRD structure already supports it.

I renamed the `inRoles` field to `roleGrants`, because I think this is more descriptive already, especially if existing roles are reconciled. Because if you know the PostgreSQL syntax for creating a role, `inRoles` and the description of the field imply that it only applies on creating the Role. But there is also reconciliation logic for this field that does not make use of `IN ROLES` at all, but uses `GRANT`.

Drawbacks:
1. The code has to be adopted to pick the `name` from the list instead of using it as-is.  
This change is the most invasive one for your PR.
2. The migration of `inRoles` isn't a plain copy-paste (which I think is fine)

mnencia replied:
The `DatabaseRole` spec deliberately inlines the same `RoleConfiguration` structure used by `.spec.managed.roles`: migrating a role is copying the stanza into a resource, and both reconcilers share the same adapter code. Changing `inRoles` into a different shape only here would fork the two APIs and create the parallel code path we both want to avoid. Also, inline roles remain supported, so there is no forced migration to lean on.

This does not block #10007. A `roleGrants` list with `admin`, `inherit` and `set` per membership can be added later as a new optional field next to `inRoles`, mutually exclusive via a validation rule, and it could land for inline and standalone roles at the same time. Deprecating `inRoles` afterwards remains an option rather than a requirement.

You have a point that the field is not only used at creation time: membership changes on an existing role are applied with `GRANT` and `REVOKE`. We amended the field description accordingly.

jonasbadstuebner replied:
Wouldn’t implementing the feature I suggested also benefit from reusing the adapter code anyway?
If inline roles remain supported, having two code paths is unavoidable.
Does your code for the `DatabaseRole` CRD handle resolving the order in which roles have to be created so you don’t run into a race condition with `IN ROLES`?

jmealo commented:
Let me know if you need help testing this. Looking forward to this! 🥳  (If we add #6043 self-service databases for app teams can be a thing out of box with some light VAP!).

leonardoce committed:
chore: avoid using SchemeBuilder

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>

mnencia committed:
test: adapt e2e to relocated test helpers

Current main restructured the e2e helper packages after this branch was
written. AssertCreateCluster now lives in tests/internal/asserts/cluster
with env and timeouts as leading arguments, CreateResourceFromFile in
tests/internal/resources, and AssertConnection together with
QueryMatchExpectationPredicate in tests/internal/asserts/postgres. This
re-points the declarative role e2e to those packages and restores the
roleExistsQuery and boolPGOutput helpers that main removed from
asserts_test.go.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia commented:
/test

mnencia committed:
fix: unblock deletion of shadowed and replica-cluster roles

The apply-only gates in shouldReconcile (role already managed inline,
replica cluster) ran before handleDeletion, so deleting a DatabaseRole
shadowed by an inline managed.roles entry, or one on a replica cluster,
left it stuck in Terminating forever, even with reclaimPolicy: retain.

Deletion now proceeds once the instance/timing gates pass, and DROP ROLE
is issued only for a role this cluster owns; a shadowed or replica role
only has its finalizer released. A missing cluster is left to the
operator, which strips the finalizers during cluster deletion.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: make role deletion idempotent on retry

roles.Delete issued a bare DROP ROLE, so a retry after the role was
already gone (for example when the finalizer update failed after a
successful drop) errored with "role does not exist" and left the
DatabaseRole stuck in Terminating. Use DROP ROLE IF EXISTS so deletion
is idempotent; this also makes the inline managed.roles ensure: absent
path tolerant of an already-absent role.

Also drop the stale "not working" TODO from the reclaim-policy docs: a
delete policy correctly keeps the resource in Terminating, retrying,
until a role's owned objects are reassigned or dropped.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
test: unit-test the instance-side reconciler

Cover the reconciler logic that previously had no unit tests and that
the two recent deletion fixes touched: shouldDropRole (drop only a role
this cluster owns), shouldReconcile (instance/timing gates plus the
deletion path that must bypass the shadowed/replica apply gates),
isAlreadyReconciled, and handleDeletion (drop an owned role, skip the
drop for shadowed or replica roles, always release the finalizer).

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: document role deletion and the adoption caveat

Add a "Removing a role" section: a role created through a DatabaseRole is
dropped via roleReclaimPolicy on resource deletion, while a pre-existing
role is removed through inline managed roles or DROP ROLE. Warn that
pointing a DatabaseRole at an existing role adopts and alters it.

Also drop the inert "ensure: present" line from the DatabaseRole example
and sample, since ensure has no effect on a DatabaseRole.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: report replica-cluster roles as Unknown, not Failed

On a replica cluster a role is owned by the primary cluster, so this
instance cannot apply it. It was marked Applied=false, which reads as a
spurious failure to fleet monitoring. Report Applied=nil ("unknown")
instead, matching the Database, Publication and Subscription reconcilers.
A role shadowed by an inline managed.roles entry stays a failure, since
that is a genuine misconfiguration to resolve.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
refactor: drop the unread transaction ID from status

The instance reconciler recorded the role's pg_authid transaction ID
in status.passwordState but never read it back. The inline
managed.roles runnable uses that value to re-assert the password when
it detects an out-of-band ALTER ROLE, but the DatabaseRole CRD is
level-triggered like the sibling Database, Publication and Subscription
objects: it reconciles only when the Kubernetes object or its
referenced password Secret changes, not on database-side drift.

Keeping the field implied a self-heal capability the CRD does not
provide. Track only the password secret's resource version, the
genuine signal that the referenced object changed.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
test: cover the operator-side password handshake

The operator records a referenced Secret's resource version in the
PasswordSecretChange condition, which is how the instance manager
learns the password changed without watching Secrets itself. That
writer side had no coverage; exercise the condition lifecycle and the
secret-to-role mapping.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: stop masking the role-list error during RBAC reconcile

When regenerating the RBAC Role granting access to the secrets
referenced by DatabaseRoles, a transient error listing the
DatabaseRoles fell through to createRole. The Role already exists
there, so the Create returns AlreadyExists and is swallowed, hiding
the listing error behind a misleading CreatingRole event and leaving
the rules unrefreshed. Return the error so the reconcile requeues.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: use a distinct role in the first sample

The first example pointed a DatabaseRole at the cluster's bootstrap
application owner. Use a separate role so the sample does not model
managing a role whose credentials the cluster maintains.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: stop leaking the secret username into status

The username-mismatch error interpolated the username read from the
password Secret. That error is surfaced in the role status, which a
wider audience can read than the Secret itself. Identify the secret
by name and the expected role instead, without echoing its contents.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: patch the status condition instead of replacing it

The operator and the instance manager both write the DatabaseRole
status, owning disjoint fields: the operator the PasswordSecretChange
condition, the instance the reconciliation result. The operator used
Status().Update, which carries the resourceVersion and conflicts when
the instance patches concurrently. Use a MergeFrom patch, matching the
instance side, so the two writers no longer race.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
refactor: share one RoleConfiguration to DatabaseRole adapter

The CRD reconciler and the inline managed-roles path each carried a
near-identical struct converting a RoleConfiguration into a
DatabaseRole, so a new field had to be mapped in both. Extract a single
DatabaseRoleFromConfiguration in the roles package and have both call
it.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
refactor: drop getters that shadow RoleConfiguration

DatabaseRoleSpec embeds RoleConfiguration, which already provides
GetRoleSecretName, GetRoleName and ShouldDisablePassword. The
re-declarations on DatabaseRoleSpec were identical and only shadowed
the promoted methods, so remove them.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
refactor: set status through the api-level helpers

The three reconciliation outcomes hand-inlined the status mutations.
Route them through SetAsApplied, SetAsFailed and a new SetAsUnknown on
the DatabaseRole type, matching the sibling Database, Publication and
Subscription reconcilers. The Patch and requeue stay in the reconciler.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia commented:
/test

gbartolini committed:
fix: spellcheck

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>

mnencia committed:
refactor: name the DatabaseRole controllers after the resource

The operator controller was Named("role"), a leftover from before the
Role CRD became DatabaseRole, and the instance controller was
Named("instance-role-reconciler"), off the instance-<resource> pattern.
Rename them to database-role and instance-database-role to match the
sibling controllers; the names only surface as controller labels in
logs and metrics.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
test: cover role deletion when the drop fails

handleDeletion released the finalizer only after a successful drop, but
the failure branch (DROP ROLE errors, e.g. the role still owns objects)
had no coverage. Assert that a failed drop keeps the finalizer, reports
the error in the status and requeues, so the role is never orphaned.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
test: cover the validUntil reset to infinity

When a role already has a validUntil and the spec drops it, the operator
translates the nil into VALID UNTIL 'infinity', because PostgreSQL cannot
restore a NULL once set. That translation had no unit coverage; assert
both the infinity reset and that it stays unset otherwise.

Assisted-by: Claude
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: never drop a role the DatabaseRole never reconciled

A conflicting duplicate DatabaseRole acquires the finalizer before the
exclusivity check blocks it, so deleting it with roleReclaimPolicy:
delete dropped the role owned by the surviving object. Gate the drop on
a recorded reconciliation, the same signal the exclusivity check uses.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: declare DatabaseRole in the OLM manifests

The OLM bundle ships every CRD from config/crd, so a CRD missing from
the CSV's owned list fails bundle validation. Add the owned-CRD entry
and the alm-examples sample, as done for the sibling CRDs.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: surface inline managed.roles takeover of an applied DatabaseRole

Once a DatabaseRole reached steady state, adding the same role to
cluster.spec.managed.roles was never noticed: the reconciler exited
through the already-reconciled gate before the shadow check and had no
Cluster watch to wake it. The CR kept reporting Applied=true while the
inline synchronizer owned the role. Detect the takeover, mark the CR
failed, and void the recorded reconciliation so it can adopt the role
back when the inline entry is removed; watch the Cluster spec so the
transition is prompt.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: document the reload label for DatabaseRole password rotation

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: state that DatabaseRole does not revert database-side changes

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: state the RBAC implications of DatabaseRole write access

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
chore: drop the unused MarkPasswordAsIgnored helper

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: polish the role management page

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
chore: list declarative-database-roles in the feature type docs

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
chore: correct a stale comment about the observed generation reset

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
refactor: rename DatabaseRoleState to DatabaseRoleStatus

Match the *Status naming of the sibling Database, Publication, and
Subscription resources before the type is frozen by the release.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: keep the original error when a status patch fails

The status helpers returned only the patch error, dropping the
reconcile error that was being reported; wrap both like the sibling
Database controller does.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: ignore DatabaseRole status changes in the cluster controller

Every status patch on a DatabaseRole triggered a full Cluster
reconciliation, while the watch only exists to refresh the instance
RBAC from the spec.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: include DatabaseRole secrets in the initial instance Role

The creation path passed no DatabaseRoles to specs.CreateRole, so on
first boot the instance manager briefly lacked permission to read the
role password secrets until the next reconciliation patched the Role.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
test: assert the surviving role outlives the conflicting CR

The conflict e2e deleted both objects without checking PostgreSQL,
so a drop wrongly issued by the losing CR went unnoticed behind the
DROP ROLE IF EXISTS idempotency.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia commented:
/test

mnencia committed:
docs: clarify how inRoles changes are applied to existing roles

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

armru committed:
fix: align the DatabaseRole schema with the sibling CRDs

Require spec and metadata like the other declarative CRDs, add the
genclient and storageversion markers, and use the same printer-column
description for the message field.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix: rename roleReclaimPolicy to databaseRoleReclaimPolicy

The sibling declarative CRDs prefix the reclaim-policy field with the
resource kind (databaseReclaimPolicy, publicationReclaimPolicy,
subscriptionReclaimPolicy); follow the same convention before the field
is frozen by a release.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
refactor: pass RoleConfiguration directly to the password helpers

The passwordManager interface had a single underlying implementation:
both call sites embed RoleConfiguration and can hand it over as-is.
Drop the interface together with the two getters that existed only to
satisfy it.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
refactor: reuse getClusterFromInstance in the DatabaseRole reconciler

The sibling reconcilers fetch their Cluster through the shared helper;
drop the hand-rolled duplicate.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: rename SetAsApplied to SetAsReady

Match the naming used by the sibling declarative resources, and explain
why the reconciler patches the status instead of going through the
shared markAs helpers: the operator and the instance manager both write
to the status, so a full update would clobber the other writer.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: drop the watch verb from the databaseroles/status grant

The instance manager watches the main resource, never the status
subresource; the sibling status grants carry get, patch and update
only.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: code-style cleanups in the DatabaseRole paths

Use index-based loops over struct slices per repo convention, drop a
no-op continue, and give the DatabaseRole field-index constant a name
that states what it indexes.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
docs: clarify adoption effects, the password condition, and capability levels

Spell out that adopting an existing role forces every attribute to the
manifest values, revoking memberships missing from inRoles and resetting
omitted attributes to their defaults. Explain the PasswordSecretChange
condition shown in the status. State that Database and DatabaseRole
objects are user-defined and reconciled by the operator, not created by
it.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: update the spellcheck wordlist for databaseRoleReclaimPolicy

Follow-up to the field rename: add the new identifier and drop the
entries that only existed for the old one.

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

mnencia committed:
fix: re-evaluate an applied DatabaseRole after the cluster is demoted

Once a DatabaseRole was applied, the already-reconciled early exit in
shouldReconcile returned before the replica gate, so a cluster demoted
to a replica kept the role reporting a stale Applied=true. Re-check the
replica condition on the designated primary and void the recorded
reconciliation, so the status reports Unknown and the role re-applies
after a promotion, matching the inline-takeover handling already there.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
fix: align the DatabaseRole cluster-immutability rule with the sibling CRDs

The cluster reference was already immutable, enforced by a spec-level
CEL rule. Express it the same way as the Database, Publication, and
Subscription CRDs instead: a field-level marker on ClusterRef carrying
the shared "cluster reference is immutable after creation" message.
Behavior is unchanged; only the rule placement and the message wording
now match the siblings. The name rule stays at the spec level because
the field is inherited from the shared RoleConfiguration.

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia commented:
/test

gbartolini committed:
docs: final review

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
